### PR TITLE
[MarkDown]: Updated handling of escape chars

### PR DESCRIPTION
**Pull request by jwoo-msft**
# Description

Updated handling of escape chars

```json

"example text \\*hello\\*"

```

The above string should renders

example text \\*hello\\*

however, it was rendered as below
example text \\\*hello\\\*

This pr fixes the issue.

Updated some of unit tests

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/7489)

### CODE REVIEW

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp [hunk 1]
```diff
@@ -249,13 +249,13 @@ namespace AdaptiveCardsSharedModelUnitTest
...
             MarkDownParser parser("foo *\\**");
-            Assert::AreEqual<std::string>("<p>foo <em>*</em></p>", parser.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <em>\\</em>*</p>", parser.TransformToHtml());
 
             MarkDownParser parser2("foo **\\***");
-            Assert::AreEqual<std::string>("<p>foo <strong>*</strong></p>", parser2.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <strong>\\</strong>*</p>", parser2.TransformToHtml());
 
             MarkDownParser parser3("foo __\\___");
-            Assert::AreEqual<std::string>("<p>foo <strong>_</strong></p>", parser3.TransformToHtml());
```
Comment (jwoo-msft) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:258]: I used Commonmark's editor to generate the correct output
```
<p>foo <em>\</em>*
foo <strong>\</strong>*
foo <strong>\</strong>_</p>
```
escaped chars within the double quotes are single characters. so \\\\ within double quotes is a single \\ in memory and outputted as shown in the above example.

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp [hunk 2]
```diff
@@ -433,7 +433,7 @@ namespace AdaptiveCardsSharedModelUnitTest
 
         TEST_METHOD(LinkBasicValidationTest_ValidLinkTestWithEscapedDelimiters)
         {
-            MarkDownParser parser("[[cool link!]](https://contoso.com/New%20Document%20(1\\).docx)");
```
Comment (jwoo-msft) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:436]: inner parenthesis is correctly handled after [PR](https://github.com/microsoft/AdaptiveCards/pull/6922) and is no longer needed to escape it.

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp [hunk 3]
```diff
@@ -249,13 +249,13 @@ namespace AdaptiveCardsSharedModelUnitTest
...
-            Assert::AreEqual<std::string>("<p>foo <em>*</em></p>", parser.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <em>\\</em>*</p>", parser.TransformToHtml());
 
             MarkDownParser parser2("foo **\\***");
-            Assert::AreEqual<std::string>("<p>foo <strong>*</strong></p>", parser2.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <strong>\\</strong>*</p>", parser2.TransformToHtml());
 
             MarkDownParser parser3("foo __\\___");
-            Assert::AreEqual<std::string>("<p>foo <strong>_</strong></p>", parser3.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <strong>\\</strong>_</p>", parser3.TransformToHtml());
```
Comment (beervoley) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:258]: @jwoo-msft @paulcam206 should we use [raw string literal](https://en.cppreference.com/w/cpp/language/string_literal) here?:)

Just for example:
```suggestion
            Assert::AreEqual<std::string>(R"<p>foo <strong>(\)</strong>_</p>", parser3.TransformToHtml());
```
We can also do `R"(raw_string)"` for more readability :)

Comment (jwoo-msft) [reply source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:258]: @beervoley 
it looks good to me.

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp [hunk 4]
```diff
@@ -249,13 +249,13 @@ namespace AdaptiveCardsSharedModelUnitTest
         TEST_METHOD(Rule11_12Test_EscapeTest)
         {
             MarkDownParser parser("foo *\\**");
-            Assert::AreEqual<std::string>("<p>foo <em>*</em></p>", parser.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <em>\\</em>*</p>", parser.TransformToHtml());
 
             MarkDownParser parser2("foo **\\***");
-            Assert::AreEqual<std::string>("<p>foo <strong>*</strong></p>", parser2.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <strong>\\</strong>*</p>", parser2.TransformToHtml());
```
Comment (jwoo-msft) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:255]: ```suggestion```
            Assert::AreEqual<std::string>(R"<p>foo <strong>\</strong>*</p>", parser2.TransformToHtml());
```

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp [hunk 5]
```diff
@@ -249,13 +249,13 @@ namespace AdaptiveCardsSharedModelUnitTest
         TEST_METHOD(Rule11_12Test_EscapeTest)
         {
             MarkDownParser parser("foo *\\**");
-            Assert::AreEqual<std::string>("<p>foo <em>*</em></p>", parser.TransformToHtml());
+            Assert::AreEqual<std::string>("<p>foo <em>\\</em>*</p>", parser.TransformToHtml());
```
Comment (jwoo-msft) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/MarkDownUnitTest.cpp:252]:
```suggestions
            Assert::AreEqual<std::string>(R"<p>foo <em>\</em>*</p>", parser.TransformToHtml());
```

**Pull request closed without merging**